### PR TITLE
MRG Layout remake

### PR DESCRIPTION
**Pull request by jaeilepp**
Though it turned out not to be the mess I expected, I decided to make a separate PR for this. This here would also be a good opportunity to design the layout so that it looks good. Now the layout uses pixels instead of inches so in theory it should work on all screen sizes and platforms.
Should there be a 'help' button somewhere that would open a pop up window to view all the key combinations etc. (discussed in https://github.com/mne-tools/mne-python/pull/2206)? Are there other buttons we might need?

### CODE REVIEW

**Comment by agramfort** [comment]
start with help button. and please share screenshots.

**Comment by jaeilepp** [comment]
The point of the remake was to make it so that it looks the same on all platforms. The actual look of the layout shouldn't have changed much (apart from the title being larger). So the idea to make this a separate PR was to test this on all platforms and see if it turns out ok. I can test this on virtual box on my laptop (which has really weird screen settings) once I get home.
screenshot from 2015-06-18 13 57 46

**Comment by agramfort** [comment]
ok cool

no time to test though

**Comment by mainakjas** [comment]
by the way, one small issue I noticed with the raw plotter. If I select the bad channel, then the channel name doesn't get grayed out so it's a bit hard to say which channel got selected. @jaeilepp can you squeeze this into one of your PRs or make a new PR to address this. Should be an easy fix ...

**Comment by Eric89GXL** [comment]
One problem is that having a non-standard DPI will break the rendering:

image


You can set the `figure.dpi` param in `matplotlibrc` if you want to try it yourself, or do the equivalent in Python space with `rcParams` or whatever it is called. Not sure if there is an easy fix. Not the end of the world if there isn't.

For me the "+", "-", and "b" keys don't work, even though pageup/pagedown/home/end do. This makes sense given that they aren't listed in `_plot_onkey` in `mne/viz/epochs.py`, are you still working on adding those?

**Comment by Eric89GXL** [comment]
Also, `'escape'` should close the epochs browsing window like it does for raw, and it should be added to the help.

File: doc/source/whats_new.rst [hunk 1]
```diff
@@ -9,6 +9,10 @@ Changelog
 ~~~~~~~~~
     - Add support for generalized M-way repeated measures ANOVA for fully balanced designs by `Denis Engemann`_
 
+    - Add epochs browser to interactively view and manipulate epochs by `Jaakko Leppakangas`_
+
+	- Add support for scaling and adjusting the number of channels/time per view by `Jaakko Leppakangas`_
```
Comment (agramfort) [inline doc/source/whats_new.rst]: indent pb

File: mne/viz/epochs.py [hunk 2]
```diff
@@ -1119,4 +1123,33 @@ def _close_event(event, params):
...
+           'down arrow : Navigate channels down\n'\
+           'up arrow : Navigate channels up\n'\
+           '- : Scale down\n'\
+           '+ : Scale up\n'\
+           '= : Scale up\n'\
+           'home : Reduce the number of epochs per view\n'\
+           'end : Increase the number of epochs per view\n'\
+           'page down : Reduce the number of channels per view\n'\
+           'page up : Increase the number of channels per view\n'\
+           'b : Toggle butterfly plot on/off\n'
```
Comment (agramfort) [inline mne/viz/epochs.py]: butterfly does not work on this branch for me.

**Comment by agramfort** [comment]
you'll need to rebase too.

**Comment by jaeilepp** [comment]
I guess it's best to finish the butterfly and continue with this when it's merged.

**Comment by agramfort** [comment]
sounds good

**Comment by jaeilepp** [comment]
Update:
Added a dialog for adjusting the number of epochs/channels per view as suggested in https://github.com/mne-tools/mne-python/issues/2213.
screenshot from 2015-06-24 15 04 04

**Comment by teonlamont** [comment]
this was missed in the last pr: could you a command line message to remind/notify the user of the drop channel similar to the drop epochs message. would be very useful.

**Comment by jaeilepp** [comment]
Well the channels aren't actually dropped. They're just added to the bads list like in raw.plot.

**Comment by teonlamont** [comment]
@mainakjas, @yousrabk, and I were discussing having an interactive way of selecting channels. In my system acquisition software we have plot of the channel montage and you can select a set of sensors this way. we could also include some buttons for pre-defined ROIs, like the ones given in mne_browse_raw (temporal, occipital quadrants, etc.)

here's a screen capture of the window:

montage_view

**Comment by teonlamont** [comment]
sorry, yeah, a message to notify which channels are marked in `info['bads']`

**Comment by jaeilepp** [comment]
There is already an 'order' param in plot_raw that allows the user to define the order of the shown channels.  Don't know if it's used much.

**Comment by mainakjas** [comment]
In the help, can you align all the colons ? And color the keyboard shortcut red / yellow or make it bold. See whatever looks good. Just some aesthetics to make it look professional :)

**Comment by mainakjas** [comment]
More help dialog related comments:

- put + and = on the same line separated by the word 'or`
- can you use the unicode to show the arrow keys: https://en.wikipedia.org/wiki/Arrow_%28symbol%29 ?
- close mne_browse_epochs -> Quit
- f11 -> F11 (make the f capital)

**Comment by mainakjas** [comment]
Some comments about the viewport dialog box

- the title of the dialog box is not inclusive of the check boxes
- the checkboxes update immediately in the plot but the viewport dimensions don't update until you press the update button. Can you make it consistent? Actually, it would be better to have no update button if possible.
- T0 visible => zeroline
- Y labels visible => channel names

Should we also have a checkbox for event-id and epoch-id for consistency?

**Comment by mainakjas** [comment]
In the help dialog, can you right justify the bold stuff (like here for example: https://inkscape.org/en/doc/keys046.html). Also can we change the sentences at the bottom to look like the table of keyboard shortcuts? Something like:

click epoch -> mark bad epoch
click channel -> mark bad channel
right click -> vertical line at a time instant
middle click -> show channel name (butterfly plot)

Maybe put a small heading which says `mouse controls` or something like that?

**Comment by agramfort** [comment]
Screen shot ?

**Comment by jaeilepp** [comment]
> the checkboxes update immediately in the plot but the viewport dimensions don't update until you press the update button. Can you make it consistent? Actually, it would be better to have no update button if possible.

This is a usability thing. If you change the viewport dimensions, it takes a while to update, which is not the case with the labels and vertical lines. Imagine trying to adjust the viewport to show every epoch and channel and not hitting the end of the slider immediately.

**Comment by jaeilepp** [comment]
screenshot from 2015-06-26 12 09 04

**Comment by jaeilepp** [comment]
The layout remake evolved a bit from the original idea. I think now the layout should be somewhat same for all platforms. At least it seems to work well for all platforms I've tried. Ready for review.

**Comment by agramfort** [comment]
I have crashes on mac os x when toggling the buttons in settings.

**Comment by jaeilepp** [comment]
Any error messages?

**Comment by agramfort** [comment]
```
  File "/Users/alex/anaconda/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 1887, in button_press_event
    self.callbacks.process(s, mouseevent)
  File "/Users/alex/anaconda/lib/python2.7/site-packages/matplotlib/cbook.py", line 540, in process
    proxy(*args, **kwargs)
  File "/Users/alex/anaconda/lib/python2.7/site-packages/matplotlib/cbook.py", line 415, in __call__
    return mtd(*args, **kwargs)
  File "/Users/alex/anaconda/lib/python2.7/site-packages/matplotlib/widgets.py", line 558, in _clicked
    if (t.get_window_extent().contains(event.x, event.y) or
  File "/Users/alex/anaconda/lib/python2.7/site-packages/matplotlib/text.py", line 798, in get_window_extent
    bbox, info, descent = self._get_layout(self._renderer)
  File "/Users/alex/anaconda/lib/python2.7/site-packages/matplotlib/text.py", line 311, in _get_layout
    ismath=False)
  File "/Users/alex/anaconda/lib/python2.7/site-packages/matplotlib/backends/backend_macosx.py", line 166, in get_text_width_height_descent
    six.text_type(s), family, size, weight, style)
RuntimeError: CGContextRef is NULL

If you suspect this is an IPython bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@scipy.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True
```

**Comment by Eric89GXL** [comment]
@jaeilepp look here:

https://github.com/mne-tools/mne-python/blob/master/mne/viz/raw.py#L322

I had to do something stupid to the raw code to prevent those errors, namely making sure the proj fig is always the last one drawn. Maybe it would fix this, too.

**Comment by jaeilepp** [comment]
I added the lines of code @Eric89GXL suggested in a couple of places. If that doesn't fix it, there might be a problem because now we have two dialogs with check boxes. If they're the ones causing the trouble we might have to find another way of toggling the labels and such. Unless someone has a better idea...

**Comment by teonlamont** [comment]
esc -> Esc

**Comment by teonlamont** [comment]
actually, the key names should probably follow the IBM naming convention: https://en.wikipedia.org/wiki/IBM_PC_keyboard

**Comment by teonlamont** [comment]
any reasoning that help isn't `h` instead of `?`

**Comment by agramfort** [comment]
gmail convention. I am fine if h works too but ? is fine with me.

**Comment by teonlamont** [comment]
oh yeah, @mainakjas just showed me that. i had no clue you had so many shortcuts in gmail. the only ones I know are from accidentally pressing a mapped key

**Comment by agramfort** [comment]
I still see crashes :( but it does not block the GUI

I would be fine with merging as it is and fixing during the sprint. If so just open an issue when merged.

ready to merge ?

**Comment by jaeilepp** [comment]
> I would be fine with merging as it is and fixing during the sprint. If so just open an issue when merged.

Yeah, I'm okay with that.
Ready for my part.

**Comment by jaeilepp** [comment]
So shall we merge this or should we wait for the sprint? I'm asking because the diff to other PRs is becoming quite big.

**Comment by dengemann** [comment]
@jaeilepp can you rebase? I can try it then.

**Comment by jaeilepp** [comment]
Done.

File: doc/source/whats_new.rst [hunk 3]
```diff
@@ -26,8 +28,11 @@ BUG
 
     - Fix FFT filter artifacts when using short windows in overlap-add by `Eric Larson`_
 
+<<<<<<< HEAD
     - Fix picking channels from forward operator could return a channel ordering different from ``info['chs']`` by `Chris Bailey`_
 
+=======
+>>>>>>> Rebase
```
Comment (agramfort) [inline doc/source/whats_new.rst]: :) :)

**Comment by agramfort** [comment]
clean up and I'll merge.

**Comment by jaeilepp** [comment]
Oops. Done.

File: mne/viz/epochs.py [hunk 4]
```diff
@@ -1353,8 +1378,196 @@ def _onpick(event, params):
...
+        params['fig_options'].show()
+        if params['fig_proj'] is not None:
+            params['fig_proj'].canvas.draw()
+    except Exception:
+        pass
+
+
+def _settings_closed(events, params):
+    """Function to handle close event from settings dialog."""
+    params['fig_options'] = None
```
Comment (agramfort) [inline mne/viz/epochs.py]: missing empty line

**Comment by agramfort** [comment]
Tell me travis is happy. I'll merge.

**Comment by jaeilepp** [comment]
Oops again. There was an empty line at the end it just doesn't show it here and I didn't notice it either.

**Comment by jaeilepp** [comment]
Okay, travis is happy.

**Comment by agramfort** [comment]
Thanks !

**Pull request merged**